constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,63 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+// AdAAddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// Both an object and an array can have values not described in the schema.
+	// A blank Type could be anything, including an object or array, so we should
+	// treat it the same as those types.
+	if sch.Type == "" || sch.Type == "object" || sch.Type == "array" {
+		trueBool := true
+		sch.XPreserveUnknownFields = &trueBool
```

Comment:
If x-preserve-unknown-fields is explicitly set to `false`, we probably want to preserve that (b/c it's a pointer I'm hoping we can tell the difference between undefined and intentionally false)

Also, no need to set XPreserveUnknown for type "array", there we just need to make sure sch.Items is structural if it's an Array.

Reply:
That makes sense, will update.